internal/updater/extract.go

```diff
@@ -23,55 +24,94 @@ func extractFile(buf []byte, filename, dest string) error {
 		mode = fi.Mode()
 	}
 
+	if err := os.Remove(dest); err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("unable to remove destination file: %q", err)
+		}
+	}
+
+	dfh, err := os.OpenFile(dest, os.O_WRONLY|os.O_CREATE|os.O_EXCL, mode)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to open file: %s", dest)
+	}
+	defer func() {
+		_ = dfh.Close()
+	}()
+
 	var rd io.Reader = bytes.NewReader(buf)
 	switch filepath.Ext(filename) {
 	case ".gz":
 		gzr, err := gzip.NewReader(rd)
 		if err != nil {
 			return err
 		}
-		rd = gzr
+		return extractTar(gzr, dfh, dest)
 	case ".bz2":
-		rd = bzip2.NewReader(rd)
+		return extractTar(bzip2.NewReader(rd), dfh, dest)
 	case ".zip":
-		return fmt.Errorf("zip archives are not supported, yet")
+		return extractZip(buf, dfh, dest)
+	default:
+		return fmt.Errorf("unsupported")
 	}
+}
 
-	if err := os.Remove(dest); err != nil {
-		if !os.IsNotExist(err) {
-			return fmt.Errorf("unable to remove destination file: %q", err)
-		}
+func extractZip(buf []byte, dfh io.WriteCloser, dest string) error {
+	zrd, err := zip.NewReader(bytes.NewReader(buf), int64(len(buf)))
+	if err != nil {
+		return err
 	}
 
-	dfh, err := os.OpenFile(dest, os.O_WRONLY|os.O_CREATE|os.O_EXCL, mode)
-	if err != nil {
-		return errors.Wrapf(err, "Failed to open file: %s", dest)
+	for i := 0; i < len(zrd.File); i++ {
+		if zrd.File[i].Name != "gopass.exe" {
+			continue
+		}
+
+		file, err := zrd.File[i].Open()
+		if err != nil {
+			return errors.Wrapf(err, "failed to read from zip file")
+		}
+
+		n, err := io.Copy(dfh, file)
+		if err != nil {
+			dfh.Close()
```

Comment:
The `defer`ed Close above would return an error finding `dfh` is already closed, but we don't check the return value of all the Close anyway... Not sure what's best.

Reply:
Yes, indeed. Doing this correctly under all circumstances is surprisingly difficult.
I really wonder if it's worth the hassle here.